### PR TITLE
Submit feedback form via Javascript

### PR DESCRIPTION
**Pull request by virginiacc**
Submits feedback form via Javascript so users receive confirmation without a page load. Adds check for `result` and `message` in JSON response to form submission. Also adds animated scrolling to bring notifications into view when they are updated on long forms like the one on `/owning-a-home/help-us-improve`.

## Additions

- Javascript form submit for feedback forms
- Animated scrolling utilities
- Styles in `forms.less` that ensure legends wrap on IE

## Changes

- Generalize `EmailSignup.js` so it can be used to submit any form
- Check for `{"result":"fail"}` and error/success messaging in JSON response to form submission

## Testing
- Test feedback form submission on [/owning-a-home/](http://localhost:8000/owning-a-home), [/owning-a-home/feedback](http://localhost:8000/owning-a-home/feedback), and [/owning-a-home/help-us-improve](http://localhost:8000/owning-a-home/help-us-improve). Verify that:
  - Form validation runs before submit and displays notification for missing comment
  - Form data is correctly submitted via Javascript
  - Messaging is displayed on successful submit
  - Notification is scrolled into view if not on screen when updated
- Also test email signup submit to ensure it still works correctly
  - It will show an error on submit locally because `{result:fail}` is returned, but the form submission also fails locally when Javascript is turned off
- In IE, verify that legend on [/owning-a-home/help-us-improve](http://localhost:8000/owning-a-home/help-us-improve) wraps correctly

## Screenshots

Feedback form validation:

screen shot 2016-12-08 at 8 18 54 am

Feedback form success:

screen shot 2016-12-08 at 8 19 46 am

Email signup validation:

screen shot 2016-12-08 at 8 20 39 am

Email signup submit messaging:
screen shot 2016-12-08 at 8 21 30 am

/help-us-improve/ legend in IE11 before:
screen shot 2016-12-08 at 8 32 26 am

and after:
screen shot 2016-12-08 at 8 31 45 am

## Checklist

* [ ] Changes are limited to a single goal (no scope creep)
* [ ] Code can be automatically merged (no conflicts)
* [ ] Code follows the standards laid out in the [front end playbook](https://github.com/cfpb/front-end)
* [ ] Passes all existing automated tests
* [ ] New functions include new tests
* [ ] New functions are documented (with a description, list of inputs, and expected output)
* [ ] Placeholder code is flagged
* [ ] Visually tested in supported browsers and devices
* [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
* [ ] Reviewers requested with the [Assignee tool](https://help.github.com/articles/assigning-issues-and-pull-requests-to-other-github-users/) :arrow_right:

### CODE REVIEW

**Comment by virginiacc** [comment]
@higs4281 I just tried switching to this branch from another one and I needed to run `./setup.sh` before the JS form submission would work. Sorry, that should have been in the testing instructions!

File: cfgov/unprocessed/js/organisms/FormSubmit.js [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+   */
+  function _cacheFields() {
+    var cachedFields = {};
+    var fields = ( _formElement || {} ).elements;
+    var f = 0;
+    var len = fields.length;
+    for ( f; f < len; f++ ) {
+      var field = fields[f];
+      if ( field.name && !field.disabled &&
+         [ 'file', 'reset', 'submit', 'button' ].indexOf( field.type ) === -1 ) {
```
Comment (sebworks) [inline cfgov/unprocessed/js/organisms/FormSubmit.js]: Might make sense to store `[ 'file', 'reset', 'submit', 'button' ]` as a variable

File: cfgov/unprocessed/js/modules/util/scroll.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+ * Easing
+ * @param {Number} t current time
+ * @param {Number} b start position
+ * @param {Number} c difference between start and end position
+ * @param {Number} d duration
+ * @returns {Number} next position
+ *
+ */
+function _easeInOutQuad( t, b, c, d ) {
+  t /= d / 2;
```
Comment (sebworks) [inline cfgov/unprocessed/js/modules/util/scroll.js]: I would use meaningful variable names here.

File: cfgov/unprocessed/js/organisms/FormSubmit.js [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+  /**
+   * @returns {obj} form fields, keyed by name.
+   *   Checkboxes and radio fields are stored in array.
+   */
+  function _cacheFields() {
+    var cachedFields = {};
+    var fields = ( _formElement || {} ).elements;
+    var f = 0;
+    var len = fields.length;
+    for ( f; f < len; f++ ) {
```
Comment (sebworks) [inline cfgov/unprocessed/js/organisms/FormSubmit.js]: I would make this for loop similar to the others.

File: cfgov/unprocessed/js/organisms/FormSubmit.js [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+          try {
+            var response = JSON.parse( xhr.responseText );
+            result = response.result;
+            message = response.message;
+          } catch( err ) {
+            // ignore lack of response
+          }
+          state = result === 'fail' ? 'ERROR' : 'SUCCESS';
+        }
+        _displayNotification( _notification[state],
```
Comment (sebworks) [inline cfgov/unprocessed/js/organisms/FormSubmit.js]: Use the notification constants here `_notification.ERROR`

File: cfgov/jinja2/v1/_includes/blocks/feedback.html [hunk 5]
```diff
@@ -1,183 +1,189 @@
...
+	      class="o-form
+	             block
+	             {% if value.was_it_helpful_text %}
+	             block__bg
+	             block__border
+	             block__padded-top
+	            {% endif %}">
+	<div class="u-mb15">
+		{% if get_messages(request) %}
+			{% for message in get_messages(request) %}
```
Comment (sebworks) [inline cfgov/jinja2/v1/_includes/blocks/feedback.html]: Should be 4 spaces here.

File: cfgov/jinja2/v1/_includes/blocks/feedback.html [hunk 6]
```diff
@@ -1,183 +1,189 @@
...
+	        <div class="form-l_col
+	                    form-l_col-1
+	                    form-l_col__flush-bottom">
+	            <legend class="h4">
+	                {{ value.was_it_helpful_text }}
+	            </legend>
+	        </div>
+	        <div class="m-big-radio-button">
+	            <div class="m-big-radio-button_container">
+	                <input class="cf-input"
```
Comment (sebworks) [inline cfgov/jinja2/v1/_includes/blocks/feedback.html]: Should you be using the input atom macro?

**Comment by higs4281** [comment]
👍 smooth!

**Comment by sebworks** [comment]
Looks good, only two things: 

1. http://localhost:8000/owning-a-home/ says the comment is optional.

2. http://localhost:8000/owning-a-home/help-us-improve/ is the only form which is hidden after form submission.

**Comment by higs4281** [comment]
@sebworks 
1. Comment on that page is indeed optional. Just the yes/no buttons are required there.
2. That is the current request.

**Comment by sebworks** [comment]
@higs4281, do you get the notification after selecting a `yes / no` button?

**Comment by virginiacc** [comment]
@sebworks Yes, it looks like the comment is being erroneously required there -- I'll push a fix for that. Thanks for catching it!

**Comment by higs4281** [comment]
@sebworks good catch! I realized I was testing the live page, which allows just yes/no.

File: cfgov/jinja2/v1/_includes/blocks/feedback.html (hunk 6 above)
Comment (virginiacc) [inline cfgov/jinja2/v1/_includes/blocks/feedback.html]: @sebworks Good point -- I do think we should be using the input macros. Right now there are differences between their layout and the design for this form, though. It seems like this is something we could revisit and either standardize the form designs or make the input macros a bit more flexible to accommodate some differences in design.

**Comment by virginiacc** [comment]
@sebworks @higs4281 Just pushed https://github.com/cfpb/cfgov-refresh/pull/2588/commits/2e02d4787eb69bf6895c4c6dafae225f0693c0eb which should fix the notification issue on the comment-optional feedback forms.

**Comment by sebworks** [comment]
💯 , great work.

**Comment by higs4281** [comment]
👍 error-checkk works for me.

**Comment by higs4281** [comment]
👍

**Comment by sebworks** [comment]
👍

**Comment by virginiacc** [comment]
Thanks, @sebworks and @higs4281!

**Pull request merged**